lib/Frontend/CompilerInvocation.cpp

```diff
@@ -973,6 +645,225 @@ StringRef FrontendArgsToOptionsConverter::determineFallbackModuleName() const {
+      !llvm::sys::fs::is_directory(unprocessedOutputFlienames.back())) {
+    Opts.OutputFilenames = unprocessedOutputFlienames;
+    return false;
+  }
+  if (Opts.Inputs.isReadingFromStdin() && unprocessedOutputFlienames.empty()) {
+    Opts.OutputFilenames = unprocessedOutputFlienames;
+    return false;
+  }
+  
+  // No output filename was specified, or an output directory was specified.
+  // Determine the correct output filename.
+  
+  // Note: this should typically only be used when invoking the frontend
+  // directly, as the driver will always pass -o with an appropriate filename
+  // if output is required for the requested action.
+  
+  assert(Opts.actionProducesOutputFromFrontend() ||
+         Opts.OutputFilenames.empty());
+  assert(Opts.actionProducesOutputFromFrontend() ||
+         !Opts.actionOutputsToStdout());
+  if (unprocessedOutputFlienames.empty() &&
+      (Opts.Inputs.isReadingFromStdin() || Opts.actionOutputsToStdout())) {
+    Opts.setOutputFilenameToStdout();
+    return false;
+  }
+  return deriveOutputFilenamesFromInputsAndSuffix();
+}
+
+bool FrontendArgsToOptionsConverter::deriveOutputFilenamesFromInputsAndSuffix() {
+  const std::vector<std::string> unprocessedOutputFlienames =
+  getUnprocessedOutputFilenames();
+  {
+    const unsigned commandLineOutputFilenameCount =
+    unprocessedOutputFlienames.size();
+    const unsigned primaryFilenameCount =
+    Opts.Inputs.primaryInputFilenameCount();
+    if (Opts.Inputs.havePrimaryInputsFilenames() &&
+        primaryFilenameCount != commandLineOutputFilenameCount &&
+        commandLineOutputFilenameCount != 0) {
+      Diags.diagnose(SourceLoc(),
+                     diag::error_output_filenames_dont_match_primary_filenames,
+                     commandLineOutputFilenameCount, primaryFilenameCount);
+      return true;
+    }
+  }
+  
+  StringRef Suffix = Opts.computeSuffix();
+  std::vector<std::string> baseNames = computeBaseNamesOfOutputs();
+  for (unsigned index : indices(baseNames)) {
+    std::string outputStem = unprocessedOutputFlienames.empty()
+    ? ""
+    : unprocessedOutputFlienames[index];
+    llvm::SmallString<128> Path(outputStem);
+    llvm::sys::path::append(Path, baseNames[index]);
+    llvm::sys::path::replace_extension(Path, Suffix);
+    Opts.OutputFilenames.push_back(Path.str());
+  }
+  
+  if (Opts.OutputFilenames.empty()) {
+    if (Opts.RequestedAction != FrontendOptions::REPL &&
+        Opts.RequestedAction != FrontendOptions::Immediate &&
+        Opts.RequestedAction != FrontendOptions::NoneAction) {
+      Diags.diagnose(SourceLoc(), diag::error_no_output_filename_specified);
+      return true;
+    }
+  } else if (Opts.isOutputFileDirectory()) {
+    Diags.diagnose(SourceLoc(), diag::error_implicit_output_file_is_directory,
+                   Opts.getSingleOutputFilename());
+    return true;
+  }
+  return false;
+}
+
+std::vector<std::string> FrontendArgsToOptionsConverter::computeBaseNamesOfOutputs() {
```

Comment:
I don't think doing this in per-output batches is the way to go. I'd rather have one outer loop over input files.